include/internal/libfreenect2/packet_processor.h

```diff
@@ -29,9 +29,99 @@
 #ifndef PACKET_PROCESSOR_H_
 #define PACKET_PROCESSOR_H_
 
+#include <cstddef>
+#include "libfreenect2/threading.h"
+
 namespace libfreenect2
 {
 
+class Buffer
+{
+public:
+  unsigned char *data;
+  size_t length;
+  size_t capacity;
+};
+
+class Allocator
+{
+public:
+  virtual Buffer *allocate(size_t size) = 0;
+  virtual void free(const Buffer *b) = 0;
+  virtual ~Allocator() {}
+};
+
+class NewAllocator: public Allocator
+{
+public:
+  virtual Buffer *allocate(size_t size)
+  {
+    Buffer *b = new Buffer;
+    b->data = new unsigned char[size];
+    b->length = 0;
+    b->capacity = size;
+    return b;
+  }
+
+  virtual void free(const Buffer *b)
+  {
+    if (b != NULL)
+      delete[] b->data;
+  }
+};
+
+template <class AllocatorT>
+class PoolAllocator: public Allocator
+{
+private:
+  AllocatorT allocator;
+  Buffer *buffers[2];
+  mutex locks[2];
+public:
+  PoolAllocator() : buffers() {}
```

Comment:
you could add a no argument ctor to poolallocator which sets the pointer to a newallocator